nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
+                                   f"bid must be at least {prettify_eth_amount(minimum)}")
         receipt = self.worklock_agent.bid(checksum_address=self.checksum_address, value=value)
         return receipt
 
     def claim(self) -> dict:
 
-        # Require the Bidding window is closed
-        end = self.worklock_agent.end_date
-        error = f"Claims cannot be placed while the bidding window is still open (closes at {end})."
-        self._ensure_bidding_is_closed(message=error)
+        # Require the cancellation window is closed
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        if not self.worklock_agent.is_claiming_available():
+            raise self.ClaimError(f"Claiming is not available yet")
 
         # Ensure the claim was not already placed
         if self._has_claimed:
-            raise self.BidderError(f"Bidder {self.checksum_address} already placed a claim.")
+            raise self.ClaimError(f"Bidder {self.checksum_address} already placed a claim.")
 
         # Require an active bid
         if not self.get_deposited_eth:
-            raise self.BidderError(f"No claims available for {self.checksum_address}")
-
-        # Ensure the claim is at least large enough for min. stake
-        minimum = self.economics.minimum_allowed_locked
-        if self.available_claim < minimum:
-            raise ValueError(f"Claim is too small. Claim amount must be worth at least {NU.from_nunits(minimum)})")
+            raise self.ClaimError(f"No bids available for {self.checksum_address}")
 
         receipt = self.worklock_agent.claim(checksum_address=self.checksum_address)
         return receipt
 
     def cancel_bid(self) -> dict:
+        self._ensure_cancellation_window(ensure_closed=False)
 
         # Require an active bid
         if not self.get_deposited_eth:
             self.BidderError(f"No bids available for {self.checksum_address}")
 
-        # Ensure the claim was not already placed
-        if self._has_claimed:
-            raise self.BidderError(f"Bidder {self.checksum_address} already placed a claim.")
-
         receipt = self.worklock_agent.cancel_bid(checksum_address=self.checksum_address)
         return receipt
 
+    def _get_max_bonus_bid_from_max_stake(self) -> int:
+        """Returns maximum allowed bid calculated from maximum allowed locked tokens"""
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        bonus_eth_supply = sum(self._all_bonus_bidders.values()) if self._all_bonus_bidders else self.worklock_agent.get_bonus_eth_supply()
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_bid = max_bonus_tokens * bonus_eth_supply // bonus_worklock_supply
+        return max_bonus_bid
+
+    def get_whales(self, force_read: bool = False) -> Dict[str, int]:
+        max_bonus_bid_from_max_stake = self._get_max_bonus_bid_from_max_stake()
```

Comment:
Needs a docstring... something like: "Returns all worklock bidders over the whale threshold as a dictionary of addresses and bid values."

Reply:
✔️